sw/host/ot_certs/src/cwt.rs

```diff
@@ -0,0 +1,967 @@
+        {size_computations}
+        {input_size_checks}
+          struct CborOut cbor;
+          RETURN_IF_ERROR(cbor_write_out_init(&cbor, buffer, *inout_size));
+
+        {cbor_instructions}
+          *inout_size = CborOutSize(&cbor);
+        {output_size_checks}
+          return kErrorOk;
+        }}
+    "#};
+
+    source_template
+}
+
+impl CwtTemplate {
+    pub fn from_hjson_str(content: &str) -> Result<CwtTemplate> {
+        deser_hjson::from_str(content).context("CwtTemplate::from_hjson_str failed")
+    }
+}
+
+pub fn load_cwt_template(path: &PathBuf) -> Result<CwtTemplate> {
+    let template_content = fs::read_to_string(path)
+        .with_context(|| format!("could not load the template file {}", path.display()))?;
+
+    CwtTemplate::from_hjson_str(&template_content).with_context(|| {
+        format!(
+            "failed to parse CwtTemplate from the template file {}",
+            path.display()
+        )
+    })
+}
+
+pub fn generate_cert(from_file: &str, template: &CwtTemplate) -> Result<Codegen> {
+    let vars = collect_codegenvar(template).context("collect_codegenvar failed")?;
+
+    let structures = CodegenStructure::from_template(template, &vars)
+        .context("CodegenStructure::from_template failed")?;
+
+    let max_size = {
+        let exp = derive_size_expressions(&structures, &|var| match var.size {
+            VariableSize::ExactSize(size) => Some(size),
+            VariableSize::MaxSize(size) => Some(size),
+        })
+        .context("derive_size_expressions for maximum size failed")?;
+
+        exp.last()
+            .context("there isn't any item in the structure")?
+            .item_size()
+    };
+
+    let decls = generate_input_fields(&vars, "  ");
+
+    let source_h = generate_header(from_file, &template.name, max_size, decls);
+
+    let exact_sizes = derive_size_expressions(&structures, &|var| match var.size {
+        VariableSize::ExactSize(size) => Some(size),
+        _ => None,
+    })
+    .context("derive_size_expressions for exact size failed")?;
+
+    let size_computations = generate_size_computations(&exact_sizes, "  ")
+        .context("generate_size_computations failed")?;
+
+    let (input_size_checks, output_size_checks) = generate_size_checks(
+        &vars,
+        exact_sizes
+            .last()
+            .context("there isn't any item in the structure")?,
+        "  ",
+    )
+    .context("generate_size_checks failed")?;
+
+    let GeneratedCborInstructions {
+        constant_definitions,
+        cbor_instructions,
+    } = generate_cbor_instructions(&structures, &exact_sizes, "  ")
+        .context("generate_cbor_instructions failed")?;
+
+    let source_c = generate_source(
+        from_file,
+        &template.name,
+        &constant_definitions,
+        &size_computations,
+        &input_size_checks,
+        &output_size_checks,
+        &cbor_instructions,
+    );
+
+    // TODO: finish the unittest
```

Comment:
is there an issue tracking this?

Reply:
Sure thing, I will create the follow up issue for the unit test and the max-size updating problem later. Thanks for all the help!